### PR TITLE
AWS: centralize logic for setting encryption info

### PR DESCRIPTION
**Pull request by jackye1995**
Try to address issue in #1767 for different request types to set encryption information.

### CODE REVIEW

**Comment by danielcweeks** [comment]
This has a lot of overlap with work done in #1767.  I'd prefer to hold off on this until work there is done since it switches to using more than just get and put operations, but needs to handle encryption for mutlipart operations.

**Comment by jackye1995** [comment]
> This has a lot of overlap with work done in #1767. I'd prefer to hold off on this until work there is done since it switches to using more than just get and put operations and needs to handle encryption for mutlipart operations.

Yes, I am thinking if we can solve the issue here and rebase that PR so that PR can be more concise. Let me run some performance check tomorrow to see how much impact it has if we use dynamic method loader to do it versus the current way using function interface.

**Comment by jackye1995** [comment]
@danielcweeks I ran some tests and here is the result:
1M average with reflection: 14.720822923822924 us
1M average with function: 2.183976667976668 us
1M average with base: 2.0982427002427 us

So the current function-based approach should provide a balance between clean interface and performance. Please let me know what you think.

P.S. for reflection approach, I use `DynMethods.UnboundMethod` to find the methods, and cache the method after creation for repeated invocation.

**Comment by danielcweeks** [comment]
@jackye1995 I just pushed #1767 in order to get the progressive upload in (it was already a pretty large PR).  We can use this issue to revisit how to handle the encryption.

I'd prefer not to use the reflection approach (not necessarily due to the performance impact since the number of S3 calls is relatively small), but rather that doesn't allow for compile time checks and will be harder to reason about and maintain.

I'm open to either the function or base approaches, but maybe once this is rebased and we add the multipart versions of this it will be more clear which approach makes the most sense.

**Comment by jackye1995** [comment]
> @jackye1995 I just pushed #1767 in order to get the progressive upload in (it was already a pretty large PR). We can use this issue to revisit how to handle the encryption.
> 
> I'd prefer not to use the reflection approach (not necessarily due to the performance impact since the number of S3 calls is relatively small), but rather that doesn't allow for compile time checks and will be harder to reason about and maintain.
> 
> I'm open to either the function or base approaches, but maybe once this is rebased and we add the multipart versions of this it will be more clear which approach makes the most sense.

Sure, let me rebase it and see

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3RequestUtil.java [hunk 1]
```diff
@@ -33,87 +36,66 @@ private S3RequestUtil() {
...
+
     switch (awsProperties.s3FileIoSseType().toLowerCase(Locale.ENGLISH)) {
       case AwsProperties.S3FILEIO_SSE_TYPE_NONE:
         break;
 
       case AwsProperties.S3FILEIO_SSE_TYPE_KMS:
-        requestBuilder.serverSideEncryption(ServerSideEncryption.AWS_KMS);
-        requestBuilder.ssekmsKeyId(awsProperties.s3FileIoSseKey());
-        break;
+        if (encryptionSetter != null) {
```
Comment (danielcweeks) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3RequestUtil.java:77]: The null checks as controls feel a little awkward.  Would it make sense to switch them out for noop functions like:

**Comment by jackye1995** [comment]
@danielcweeks checks passed

**Comment by danielcweeks** [comment]
+1

**Pull request merged**